### PR TITLE
Implement structs in the frontend

### PR DESCRIPTION
**Pull request by jfecher**
This PR implements the majority of frontend functionality needed for structs in noir.

Included in this PR:
- 3 new node types and support for parsing them:
    - NoirStruct for struct definitions (currently only allowed at top-level to match functions)
    - ConstructorExpression for `MyStruct { field: expr, ... }`
    - MemberAccessExpression for `a.field`
- Name resolution for structs, including a few new error types related to duplicate fields, or otherwise incorrect fields provided for a ConstructorExpression.
- Type checking for struct types. Note also that in a member access expression like `a.field` the check for whether `a` actually has a field named `field` is deferred to type checking since that is the only time we know the type of `a`.
- An example is provided in `examples/struct`. This example is meant only to pass through to the backend where it should fail since structs are not implemented there yet. However, the test currently passes when it shouldn't.

NOT included in this PR:
- Any interpreting or codegen for struct definitions, member access expressions, or constructor expressions.

Notes:
- Unlike most other constructs, `StructType`s are stored in an `Rc` rather than an interner to avoid the need to pass around a context everywhere types are used. Doing the later would also prevent us from implementing traits like `Display for Type` since we would require a context to display struct types. I think the `Rc` approach should be more ergonomic for now.
- Name resolution, particularly def collection and import resolution, were by far the most difficult parts of the frontend to understand and extend. I wonder if these can be simplified somehow.
- This PR also adds `Cargo.lock` to the .gitignore to avoid it cluttering up every git commit when it changes

### CODE REVIEW

File: crates/noirc_frontend/src/hir/def_collector/dc_mod.rs [hunk 1]
```diff
@@ -60,27 +63,64 @@ impl<'a> ModCollector<'a> {
             unresolved_functions.push_fn(self.module_id, func_id, nf);
 
             // Add function to scope/ns of the module
-            self.def_collector.def_map.modules[self.module_id.0]
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_frontend/src/hir/def_collector/dc_mod.rs:63]: LGTM

File: crates/noirc_frontend/src/parser/parser.rs [hunk 2]
```diff
@@ -97,10 +106,21 @@ fn attribute() -> impl NoirParser<Attribute> {
     })
 }
 
+fn struct_fields() -> impl NoirParser<Vec<(Ident, Type)>> {
+    parameters(parse_type_no_field_element())
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_frontend/src/parser/parser.rs:110]: This works, although I think in the future, we may want to have priv and const, so we can differentiate between a private variable and a constant variable. Not blocking for this PR and we can open an issue for it

[the remark this answers is not in the document]
Comment (jfecher) [reply crates/noirc_frontend/src/parser/parser.rs]: I changed it to allow `priv` or `const`

**Review by kevaundray — DISMISSED** [review]

**Comment by kevaundray** [comment]
LGTM

**Review by kevaundray — APPROVED** [review]

**Pull request merged**